Fix #7562: deployment workflow

L-TChen opened the pull request:
Agda no longer requires `agda-mode` or environment variables to locate primitive modules. As such, the deployment workflow has been updated accordingly. A sample release can be found in my fork: https://github.com/L-TChen/agda/releases/tag/nightly

The new workflow builds binaries (with the flag `--optimise-heavily`) for the following platforms, as before:

1. Linux (statically linked)
2. macOS (x64)
3. macOS (arm64 / Apple silicon)
4. Windows

Each binary is tested in a clean environment to ensure it runs correctly (using `agda --version`). If the workflow is triggered by a tag or a commit to `master`, the binaries are attached to the release.

Note: I have disabled the `--enable-cluster-counting` flag on all platforms, as I’m currently unable to statically link `icu4c`.

### Remaining tasks

* [ ] Document the pre-built binaries. Should we recommend their use?
* [ ] Update the changelog.
* [ ] (Optional) Notarisation for macOS (requires an Apple Developer account; a fee waiver may be possible).

### macOS

macOS binaries are not [notarised](https://developer.apple.com/documentation/security/notarizing-macos-software-before-distribution), so users must remove the quarantine attribute before use (see [discussion](https://discourse.haskell.org/t/mac-notarization-and-ghc/5722)). This explains the following step in the sanity check:

```bash
if [[ "$MATRIX_OS" == "macos"* ]]; then
  xattr -rc Agda/agda
fi
```

L-TChen committed:
Remove `enable-cluster-couting`, scripts, and other dlls

L-TChen committed:
Check the generated binary after compilation in a new job

andreasabel reviewed (commented):
Need to run the whitespace fixer and update the PR...

L-TChen commented:
> Need to run the whitespace fixer and update the PR...

It was okay locally, and I didn't know why it failed. Now, I have rebased the branch and it passes the check. 

###  Notarisation and code signing
The documentation depends on how we treat these binaries: as a by-product of CI or something more official.  If we are able to sign these binaries, then we may distribute these binaries without causing a warning about unrecognised programs. I can get an Apple Developer account for testing the workflow first. 

We may also discuss this in a separate PR/issue.

ice1000 commented on src/github/workflows/deploy.yml:
```diff
@@ -327,9 +366,12 @@ jobs:
+        for i in artifacts/**/*; do mv "$i" "$(dirname ${i})/Agda-${sha}-$(basename ${i})"; done
         ls -R artifacts
-        gh release create 'nightly' artifacts/**/* --repo "${GITHUB_REPOSITORY}" --generate-notes --notes-start-tag v2.7.0 --prerelease --title "${{ needs.build.outputs.sha }}@master"
+        gh release create 'nightly' artifacts/**/* --repo "${GITHUB_REPOSITORY}" --generate-notes --notes-start-tag v2.7.0.1 --prerelease --title "${sha}@master"
```
Do we have to change this every time we release? 🤔

L-TChen replied:
https://github.com/marketplace/actions/get-latest-tag Maybe this can be used?

ice1000 replied:
Yeah, but does Agda have consistent tags?

andreasabel replied:
> https://github.com/marketplace/actions/get-latest-tag Maybe this can be used?

This does not specify what it means by "latest tag", so one would have to try it out.  Latest by date certainly does not work.  What we want is the tag `v<ListNat>` with the lexicographically greatest `<ListNat>`.

andreasabel replied:
> Do we have to change this every time we release? 🤔

Yes:
https://github.com/agda/agda/blob/2eba43d8f45c9d380c4e88d172d1533397544d36/notes/agda-releases.txt#L162-L165

L-TChen replied:
It appears that the following command works (at least, locally)

`git tag -l 'v*' --sort=-version:refname | grep -v -e '-rc[0-9]\+$' -e '-r[0-9]\+$' | head -n 1`

where
1. `--sort=-version:refname` sorts the list of tags in the descending order by the version numbers(*) (so `v2.10` comes before `v2.9`) and
2. the `grep` excludes release candidates. 

In the current repository, the above command gives `v2.7.0.1` as desired. 

(*) The git manual does not specify if this is exactly the lexicographical order.

andreasabel commented on .github/workflows/deploy.yml:
```diff
@@ -264,8 +214,76 @@ jobs:
+          xattr -rc Agda/agda
+        fi
+
+        Agda/agda --version
```
It would make sense to also try `agda --setup` here...

andreasabel commented on src/github/workflows/deploy.yml:
```diff
@@ -306,12 +273,84 @@ jobs:
+          xattr -rc Agda/agda
+        fi
+
+        Agda/agda --version
```
We could also try out `agda --setup` here.

L-TChen replied:
Done.

L-TChen committed:
Ditch the additional directory structure in the compressed files;
place platform-independent variables after other variables

L-TChen committed:
Retrieve the latest tag using `git tag -l`

L-TChen commented:
Shall we recommend the use of these pre-built binaries? If not, I may just drop a line in the documentation about their mere existence.

ice1000 commented:
I think they should at least be in the documentation...

L-TChen committed:
Describe the pre-built binaries in doc

Also remove the warning about melpa

L-TChen committed:
Mention the pre-built binaries

L-TChen commented:
I have updated the Agda user manual and added a description to the changelog. 

This PR should be ready to go, except some upstream problem about `nonempty-containers`.

andreasabel commented on doc/user-manual/getting-started/installation.rst:
```diff
@@ -22,7 +22,42 @@ In case of installation problems, check the section on :ref:`troubleshooting <tr
+   `GHCup installation instructions <https://www.haskell.org/ghcup/>`_ to install GHC.
+
+.. note::
+   These binaries are compiled with the :option:`optimise-heavily` flag only,
```
Would be good to add that `enable-cluster-counting` is off in these prebuilt binaries.

L-TChen replied:
Done!

andreasabel commented on doc/user-manual/getting-started/installation.rst:
```diff
@@ -193,10 +228,6 @@ by the Agda developers in the main Agda repository and offers many advanced feat
-.. Warning::
-  Installing ``agda-mode`` via ``melpa`` is discouraged.
-  It is strongly advised to install ``agda-mode`` for ``emacs`` as described below:
-
```
Why was this warning deleted?

L-TChen replied:
`agda-mode` was removed from `melpa` almost four years ago (https://github.com/melpa/melpa/pull/7743), so I think it is safe to remove the warning.

andreasabel reviewed (commented):
Thanks for the updated and the docs!

L-TChen committed:
Document the missing flag `enable-cluster-counting`

wenkokke commented:
> Shall we recommend the use of these pre-built binaries? If not, I may just drop a line in the documentation about their mere existence.

setup-agda actually uses these binaries if the user selects the nightly version. However, I'm not sure if we can recommend the nightly over other versions?

wenkokke commented:
Btw, setup-agda's bundles have nearly the same structure, except that they don't contain your scripts and in most cases are linked against and bundled with ICU.

We could fairly easily replace a lot of the intricacies of this workflow with a call to setup-agda build.

andreasabel commented:
@L-TChen : Ok to squash this or should commits be preserved?

@wenkokke wrote:
> We could fairly easily replace a lot of the intricacies of this workflow with a call to setup-agda build.

That could be interesting.  I am inclined to merge this PR first though.  Would you be willing to PR a reformulation using `setup-agda`?

L-TChen commented:
> @L-TChen : Ok to squash this or should commits be preserved?
> 
I will squash it later. (Just double check ...)
> @wenkokke wrote:
> 
> > We could fairly easily replace a lot of the intricacies of this workflow with a call to setup-agda build.
> 
> That could be interesting. I am inclined to merge this PR first though. Would you be willing to PR a reformulation using `setup-agda`?

It looks ideal. My concern is that this workflow will also be used as an example to build a deployable binary of the Agda language server for Wasm. We are using an experimental GHC and a derivative of Agda, so I will still have to maintain some variant of this workflow somewhere. Can `setup-agda` help with that? 

There are some other features I'd like to have about `setup-agda`, but we shall have another issue for this.
